Document OTel bridge (#1966)
elastic/apm-agent-dotnet: docs/api-opentelemetry.asciidoc

ifdef::env-github[]
NOTE: For the best reading experience,
please view this documentation at https://www.elastic.co/guide/en/apm/agent/dotnet[elastic.co]
endif::[]

[[opentelemetry-bridge]]
== OpenTelemetry Bridge

> NOTE: Added as experimental in v1.13.0.
> To enable it, set <<config-enable-open-telemetry-bridge, `OpentelemetryBridgeEnabled`>> to `true`.
> 
> NOTE: The OpenTelemetry bridge is supported on .NET 5 and newer versions.
> 

Reviewer: I'd prefer to use our built-in `preview::[]` as it includes legal jargon about how we support technical preview features. What about:

```suggestion
preview::[]

NOTE: This feature was added in v1.13.0. It is supported on .NET 5 and newer versions.

```

> The OpenTelemetry Bridge in the Elastic .NET APM Agent bridges OpenTelemetry spans into Elastic APM transactions and spans. The Elastic APM OpenTelemetry bridge allows one to use the vendor-neutral OpenTelemetry Tracing API to manually instrument one's code, and have the Elastic .NET APM agent handle those API calls. This allows one to use the Elastic APM agent for tracing, without any vendor lock-in from adding manual tracing using the APM agent’s own <<public-api, Public API>>.

Reviewer: ```suggestion
The OpenTelemetry Bridge in the Elastic .NET APM Agent bridges OpenTelemetry spans into Elastic APM transactions and spans. The Elastic APM OpenTelemetry bridge allows you to use the vendor-neutral OpenTelemetry Tracing API to manually instrument your code and have the Elastic .NET APM agent handle those API calls. This means you can use the Elastic APM agent for tracing, without any vendor lock-in from adding manual tracing using the APM agent’s own <<public-api, Public API>>.
```

OpenTelemetry in .NET is implemented via the https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity?view=net-5.0[Activity API].

In .NET, there is an https://opentelemetry.io/docs/instrumentation/net/shim/[OpenTelemetry shim] which is built on top of the Activity API. 

> The OpenTelemetry bridge in the Elastic .NET APM Agent targets the https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity?view=net-5.0[Activity API]. Since the https://opentelemetry.io/docs/instrumentation/net/shim/[OpenTelemetry .NET shim] builds on top of the https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity?view=net-5.0[Activity API], the shim is implicitly supported as well, although we don't directly test it, because the Activity API is the recommended API.

Reviewer: A couple of thoughts here.

1. This seems repetitive
2. Are these implementation details?

Can we move this to the end? Maybe a "how this feature works" section?

Author: I moved it into the end.

What I wanted to say here is this:
- OTel has its predefined API (like `Tracer`, `Span`, etc.) - of course .NET does not follow this :D In .NET there is this thing called `Activity` which is the equivalent of a span, so the API (especially the name of specific types, and functions) for .NET OTel is different - things are naming differently and the .NET implementation doesn't follow the naming of the spec. (I know... not nice, but that's how it is). So I wanted to say that when we talk about the OTel bridge in .NET, what we mean is that we bridge the `Activity` API (which is de-facto OTel in .NET).
- To make things more complicated there is actually an alternative API that strictly follows the spec - that's the "shim" I talk about. That's built on top of `Activity`, and this shim is not recommended to use (the doc of it literally says "use `Activity` instead"). But I wanted to point out, that we indirectly support this shim as well.

> Are these implementation details?

So I would not say this is implementation detail - I'm just trying to communicate what API we actually support by the bridge.

I named the new section at the end `Supported OpenTelemetry implementations` - I think that better describes what this is.

> This seems repetitive

I changed the text a little bit.

Reviewer: Aha! Thanks. This makes sense to me now. I like it at the end if you're okay with it there. And the rewording is great.

[float]
[[otel-getting-started]]
=== Getting started

> The OpenTelemetry bridge is part of the core agent pacakge (https://www.nuget.org/packages/Elastic.Apm[`Elastic.Apm`]), so you don't need to add any additional dependency.

Reviewer: ```suggestion
The OpenTelemetry bridge is part of the core agent package (https://www.nuget.org/packages/Elastic.Apm[`Elastic.Apm`]), so you don't need to add an additional dependency.
```

[float]
[[otel-enable-bridge]]
==== Enable the OpenTelemetry Bridge

Currently the bridge is opt-in, so you need to enable it.

If you configured the agent via environment variables, set the `ELASTIC_APM_ENABLEOPENTELEMETRYBRIDGE` environment variable to `true``.

If you configured the agent via the `appsettings.json` file, then set `ElasticApm:EnableOpenTelemetryBridge` to `true`.

[source,js]
----
{
  "ElasticApm":
    {
      "ServerUrl":  "http://myapmserver:8200",
      "SecretToken":  "apm-server-secret-token",
      "EnableOpenTelemetryBridge": true
    }
}
----

[float]
[[create-activity-source-and-spans]]
==== Create an ActivitySource and start spans

> Once you enabled the bridge, you can create OpenTelemetry spans, or in .NET terminology you can start created new activities via the activity source and the agent will bridge those spans automatically.

Reviewer: ```suggestion
After enabling the bridge, you can create OpenTelemetry spans, or in .NET terminology, you can start creating new activities via the activity source, and the agent will bridge those spans automatically.
```

[source,csharp]
----
public static void Sample()
{
	var src = new ActivitySource("Test");
	using var activity1 = src.StartActivity(nameof(Sample), ActivityKind.Server);
	Thread.Sleep(100);
	using var activity2 = src.StartActivity("foo");
	Thread.Sleep(150);
}
----

> The code-snippet above creates a span called `Sample` and a child span on `Sample` called `foo`. The bridge will create a transaction from `Sample` and a child span called `foo`.

Reviewer: ```suggestion
The code snippet above creates a span named `Sample` and a child span on `Sample` named `foo`. The bridge will create a transaction from `Sample` and a child span named `foo`.
```

[float]
[[mixing-apis]]
==== Mixing OpenTelemetry and the Elastic APM Public API

You can also mix the Activity API with the <<public-api, Public API>>, the OpenTelemetry bridge will take care of putting the spans into the right place. The advantage of this is that if you already have some libraries that you instrumented via the <<public-api, Public API>>, but going forward, you'd like to use the vendor-independent OpenTelemetry API, you don't need to replace all Public API calls in one go.

[source,csharp]
----
/// ElasticTransaction
/// -
/// ---> OTelSpan
///           -
///           ---> ElasticSpan

var src = new ActivitySource("Test");
tracer.CaptureTransaction( nameof(Sample4), "test", t =>
{
	Thread.Sleep(100);
	using (var activity = src.StartActivity("foo"))
	{
		tracer.CurrentSpan.CaptureSpan("ElasticApmSpan", "test", () => Thread.Sleep(50));
		Thread.Sleep(150);
	}
});
----

> The code-snippet above creates a transaction with the Elastic .NET APM Agent's <<public-api, Public API>>. Then it creates an activity called `foo`. This activity is going to be a child of the previously created transaction. Then finally, a span is created again by using the Elastic .NET APM Agent's <<public-api, Public API>>. This span is going to be a child span of the OpenTelemetry span.

Reviewer: ```suggestion
The code snippet above creates a transaction with the Elastic .NET APM Agent's <<public-api, Public API>>. Then it creates an activity called `foo`; this activity will be a child of the previously created transaction. Finally, a span is created again using the Elastic .NET APM Agent's <<public-api, Public API>>; this span will be a child span of the OpenTelemetry span.
```

ifdef::env-github[]
NOTE: For the best reading experience,
please view this documentation at https://www.elastic.co/guide/en/apm/agent/dotnet[elastic.co]
endif::[]

[[opentelemetry-bridge]]
== OpenTelemetry Bridge

preview::[]

NOTE: This feature was added in v1.13.0. It is supported on .NET 5 and newer versions.

The OpenTelemetry Bridge in the Elastic .NET APM Agent bridges OpenTelemetry spans into Elastic APM transactions and spans. The Elastic APM OpenTelemetry bridge allows you to use the vendor-neutral OpenTelemetry Tracing API to manually instrument your code and have the Elastic .NET APM agent handle those API calls. This means you can use the Elastic APM agent for tracing, without any vendor lock-in from adding manual tracing using the APM agent’s own <<public-api, Public API>>.

[float]
[[otel-getting-started]]
=== Getting started

The OpenTelemetry bridge is part of the core agent package (https://www.nuget.org/packages/Elastic.Apm[`Elastic.Apm`]), so you don't need to add an additional dependency.

[float]
[[otel-enable-bridge]]
==== Enable the OpenTelemetry Bridge

Currently the bridge is opt-in, so you need to enable it.

> If you configured the agent via environment variables, set the `ELASTIC_APM_ENABLEOPENTELEMETRYBRIDGE` environment variable to `true``.

Reviewer: ```suggestion
If you configured the agent via environment variables, set the `ELASTIC_APM_OPENTELEMETRY_BRIDGE_ENABLED ` environment variable to `true``.
```
